File: src/org/likexin/binarytree/InorderTraversal.java
package org.likexin.binarytree;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * http://www.lintcode.com/en/problem/binary-tree-inorder-traversal/
 *
 * @author Shingo Lee
 */
public class InorderTraversal {

  /**
   * 中序遍历：左根右，栈的特征是先入后出。
   *
   * @param root 二叉树根结点
   * @return 中序遍历集合
   */
  public List<Integer> inorderTraversal(TreeNode root) {
    Stack<TreeNode> stack = new Stack<>();
    List<Integer> result = new ArrayList<>();
    TreeNode curr = root;

    while (curr != null || !stack.empty()) {
      while (curr != null) {
        stack.add(curr);
        curr = curr.left;
      }
      stack.pop();
      result.add(curr.val);
      curr = curr.right;
    }
    return result;
  }

}

File: Python/binarytree/inorderTraversal.py
# -*- coding:UTF-8 -*-
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # solution 1: recursion
        if not root:
            return []
        ans = []
        ans.append(root.val)
        ans = self.inorderTraversal(root.left) + ans + self.inorderTraversal(root.right)
        return ans
        

        # solution 2: loop
        if not root:
            return []
        ans, stack = [], []
        p = root
        while p or stack:
            while p:
                stack.append(p)
                p = p.left
            if stack:
                temp = stack.pop()
                ans.append(temp.val)
                p = temp.right
        return ans
    

                
        
        
        